Yoast/wordpress-seo
# Add network feature toggles

## Summary

This PR can be summarized in the following changelog entry:

* Introduces a Features tab in the network admin, which allows disabling all site-specific features for the entire network.

## Relevant technical choices:

### Data

* Each of the feature variables is now added to the `wpseo_ms` option, prefixed with `allow_` so that it doesn't collide with the `wpseo` option variable names.
* The option class now supports an `$override_option_name` property. When this is set, the option with that name is fetched from the next higher level in the hierarchy. This is abstracted a bit, but essentially it means the following:
    * For regular (site) options, network options are the higher level (if multisite setup).
    * All other options currently do not have any higher level.
* Any override option vars that match a name of one of the option vars, prefixed with `allow_` will cause this option to not be changeable when they are set to false.
* The `wpseo_ms` network option is set as the `$override_option_name` for the `wpseo` option (so that the new `allow_*` network option vars prevent updates to the respective option vars accordingly).
* A very specific adjustment has been made to the `wpseo` option's class: For all feature options, their values are dynamically set to false if the respective `allow_*` network option is set to false, via a new filter callback that is hooked into both `option_wpseo` and `default_option_wpseo`. This is very specific for now, so I didn't bother abstracting this, until we know whether we need something similar again.

### UI

* A new `Yoast_Feature_Toggles` class (and related `Yoast_Feature_Toggle` class) is introduced, containing the labels, help information etc. for all features. This is only UI-related for now, therefore in a subdirectory of `/admin`. It shouldn't become more global at this point, since the regular options handling doesn't require awareness of UI things such as labels.
* `Yoast_Form` now stores the value of the current option's override option if it is set.
* It looks at the `allow_*` variables for that option (if defined) and for each input control rendered sets the `disabled` attribute in case that value is set and false.
* @hedgefield The toggles in the site admin currently do not visually indicate that they are disabled, although the attribute is present. We'd need to get this to be obvious. It would also help to somehow incorporate a brief text such as "Feature disabled by network administrator." along with every toggle that is disabled, just so that people don't wonder why they can't change the setting. I'd furthermore ask you to generally review specifically the wording of the network feature toggles screen.

## Test instructions

This PR can be tested by following these steps:

* Activate the plugin on a multisite.
* run `composer du`
* Go to the network admin, and from there to the plugin's settings screen.
* Verify there is a Features tab, and access it.
* To make testing easy, open another tab (in your browser), and in there go to the regular (site) admin, to the plugin' settings screen and the Features tab in there.
* In the site admin's Features tab, randomly set the On/Off values for each feature and hit Save Changes.
* In the network admin's Features tab, now disable a few of the features, and make sure at least one of them is a feature which is enabled in the site admin. Remember that specific one, because the rest of tests you're gonna focus on only it. Hit Save Changes (saving here should happen via AJAX, without page reload).
* Reload the browser tab in which you have the site admin's Features tab opened.
* Verify that the features you disabled in the network admin now show up as greyed-out (disabled) and that they show as Off.
* Now, open the browser's inspector, and remove the `disabled` attribute from one of the disabled toggles that was originally set to being enabled in the site admin. Hit Save Changes (this is to test setting a disabled control to false and thus changing its value although you aren't supposed to do that).
* After the reload, verify the toggle you hacked around with is still disabled and still set to false.
* In the network admin's Features tab, change that specific toggle to being enabled again, and hit Save Changes.
* Reload the browser tab in which you have the site admin's Features tab opened.
* Verify that the feature you just enabled again in the network admin is no longer greyed-out and that it is again enabled, similar to how it originally was before you disabled it in the network admin.

## Quality assurance


Fixes #8535
